src/client/fuse_ll.cc

```diff
@@ -163,6 +168,101 @@ class CephFuse::Handle {
   struct fuse_args args;
 };
 
+static bool already_mounted(const char *path)
+{
+  struct stat orig_stat;
+  struct stat parent_stat;
+  char *orig_path = NULL;
+  char *parent_path = NULL;
+  bool ret_val = false;
+  int err = 0;
+
+  do
+  {
+    orig_path = new char[strlen(path) + 1];
+    memcpy(orig_path, path, strlen(path) + 1);
+    parent_path = dirname(orig_path);
+
+    // get stat information for original path
+    err = stat(path, &orig_stat);
+    if ( -1 == err ) {
+      derr << "fuse_ll: already_mounted: stat(" << path << ") failed with error "
+	<< cpp_strerror(err) << dendl;
+      break;
+    }
+
+    // if path isn't directory, then it can't be a mountpoint.
+    if ( !(orig_stat.st_mode & S_IFDIR) ) {
+      derr << "fuse_ll: already_mounted: "
+	<< path << " is not a directory" << dendl;
+      break;
+    }
+
+    // get stat information for parent path
+    err = stat(parent_path, &parent_stat);
+    if ( -1 == err ) {
+      derr << "fuse_ll: already_mounted: stat(" << parent_path << ") failed with error "
+	<< cpp_strerror(err) << dendl;
+      break;
+    }
+
+    // if original path and parent have different device ids,
+    // then the path is a mount point
+    // or, if they refer to the same path, then it's probably
+    // the root directory '/' and therefore path is a mountpoint
+    if( orig_stat.st_dev != parent_stat.st_dev ||
+	( orig_stat.st_dev == parent_stat.st_dev &&
+	  orig_stat.st_ino == parent_stat.st_ino
+	)
+      ) {
+      ret_val = true;
+    }
```

Comment:
A better check I would think is to simply test if `orig_stat.st_dev` is a fuse device. If it is, you're done. No need to look at `parent_stat`.

Reply:
No. That would prevent you from doing nested ceph-fuse mounts.

This check is to test whether the dentry is a mountpoint at all. st_dev is dynamically assigned by the kernel for anon mounts like FUSE, so you can't easily use that to tell whether the mount is FUSE.

In principle, you could go on to do a statfs(2) on the mountpoint, and see whether the f_type was FUSE_SUPER_MAGIC, and reject the mount if it is. That might be more efficient (and reliable) than scraping the mtab, but it would prevent you from overlaying a different type of FUSE mount with ceph-fuse. Maybe that's ok...

Reply:
Using getxattr and checking ceph.client_id attribute to verify ceph fuse mount.